akka-kafka-to-sharding/src/main/scala/sample/sharding/kafka/TopicListener.scala

```diff
@@ -0,0 +1,33 @@
+package sample.sharding.kafka
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.cluster.sharding.dynamic.DynamicShardAllocation
+import akka.cluster.typed.Cluster
+import akka.kafka.ConsumerRebalanceEvent
+import akka.kafka.TopicPartitionsAssigned
+import akka.kafka.TopicPartitionsRevoked
+import akka.actor.typed.scaladsl.adapter._
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+
+object TopicListener {
+  def apply(typeKey: EntityTypeKey[_], partitionToShard: Int => String = _.toString): Behavior[ConsumerRebalanceEvent] =
+    Behaviors.setup { ctx =>
+      val shardAllocationClient = DynamicShardAllocation(ctx.system.toClassic).clientFor(typeKey.name)
```

Comment:
This is an extension that links up with the dynamic shard allocation strategy rather than exposing the internal ddata structure directly. We can provide a typed extension that takes the type key in directly